src/index.ts

```diff
@@ -0,0 +1,87 @@
+export type StackErrorJSON = {
+  code: string
+  message: string
+  metadata: Record<string, unknown>
+  name: string
+  stack: Array<StackErrorJSON>
+}
+
+export class StackError extends Error {
+  static fromJSON(json: StackErrorJSON): StackError {
+    const error = new StackError(json.code, json.message)
+    error.errorStack = (json.stack ?? []).reduceRight((stack, e) => {
+      const err = StackError.fromJSON(e)
+      err.errorStack = stack
+      return [err, ...stack]
+    }, [] as Array<StackError>)
+    error.metadata = json.metadata ?? {}
+    error.name = json.name ?? 'StackError'
+    return error
+  }
+
+  code: string
+  errorStack: Array<StackError>
+  metadata: Record<string, unknown> = {}
+  name = 'StackError'
+
+  constructor(code: string, message: string, parentError?: StackError) {
+    super(message)
+    Object.setPrototypeOf(this, StackError.prototype)
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, StackError)
+    }
+
+    this.code = code
+    this.errorStack = parentError ? parentError.toErrorStack() : []
+  }
+
+  toErrorStack(): Array<StackError> {
+    return [this, ...this.errorStack]
+  }
+
+  toJSON(withStack = true): StackErrorJSON {
+    return {
+      code: this.code,
+      message: this.message,
+      metadata: this.metadata,
+      name: this.name,
+      stack: withStack ? this.errorStack.map(e => e.toJSON(false)) : []
+    }
+  }
+
+  toString(): string {
+    return `[${this.code}] ${this.message}`
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function createNamespaceError(namespace: string, metadata: Record<string, unknown> = {}) {
+  return class NamespaceError extends StackError {
+    constructor(code: string | number, message: string, parentError?: StackError) {
+      super(`${namespace}${code}`, message, parentError)
+      Object.setPrototypeOf(this, NamespaceError.prototype)
+      if (Error.captureStackTrace) {
+        Error.captureStackTrace(this, NamespaceError)
+      }
+
+      this.metadata = metadata
+    }
+  }
+}
+
+export function assert(condition: boolean, message = 'Assertion failed'): asserts condition {
+  if (!condition) {
+    throw new Error(message)
+  }
+}
+
+export function assertAs<T extends typeof StackError>(
+  condition: boolean,
```

Comment:
I think we should discourage asserting a raw condition whenever possible.  The problem with this is if the condition fails we wind up with a message like `"assert failed: false"`.  This is why I prefer assertion functions like `assert.equal`, `assert.notEqual`, etc as you can print the arguments given and make the error messages more descriptive

Reply:
Yes makes sense, I wanted to keep this lib limited to the core functionalities as it will likely be used in many packages to avoid making it heavier than it needs.
Maybe we could create an assertions utilities library using this base one, with a factory such as:
```ts
function createAssert(ErrorClass) {
  function assert(condition: boolean, code: string | number, message: string) {
    return assertAs(condition, ErrorClass, code, message)
  }

  assert.equal = (a, b) => {
    return assert(a === b, 11, `${a} must be equal to ${b}`)
  }
  assert.notEqual = (a, b) => {
    return assert(a !== b, 12, `${a} must not be equal to ${b}`)
  }
  assert.deepEqual = ...
  ...

  return assert
}
```

That could be used such as:
```ts
const LibError = createNamespaceError('LIB')
const assert = createAssert(LibError)
asset.equal(a, b)
```

Reply:
Interesting idea!  I think my concern with the structure as presented is that it becomes harder to assert with different codes within the same file.  Your example gives `assert.equal` and `assert.notEqual` different consistent error codes, but that isn't how I imagined error codes being used.  I imagined we'd have codes with semantic meaning more like `FileNotFound`, `HostUnavailable`, `IncompatibleSchema` etc.  So for that we'd need something like:
`assert.equal = function(a: any, b: any, code: ErrorCode, contextMessage: string) { /* ... */ }`
Where `ErrorCode` is an enum containing all of our semantically meaningful error codes and `contextMessage` is a string that gets combined with the more generic `"${a} must be equal to ${b}"` message to provide more context to the specific failure.  So this could be used in a way that looks something like:

```
function doStuffWithRecord(cid: CID) {
  // ...
  const record = this.findRecord(cid)
  assert.notEqual(record, null, ErrorCodes.RecordNotFound, "No record found with the provided CID")
  // ...
}
```

Reply:
Actually I think what you wrote and what I wrote aren't mutually exclusive.  We can just change the assert.equal and assert.notEqual functions you wrote to have error code and message arguments.  I like this approach and would love to see it included in this CR so we can have a complete picture of how this actually gets used in practice

Reply:
I don't think this `createAssert()` function should live in this library, on the contrary it should be up to each package to use what suits them best.

For example to support Ceramic's use cases we could have a `ceramic-errors` package with
```ts
class ProtocolError extends createNamespaceError('CRP') {
  toAPI() { ... }
}

function assertProtocolVersion(version) {
  return assertAs(version === PROTOCOL_VERSION, ProtocolError, 1, `Invalid protocol version: expected ${PROTOCOL_VERSION}, got ${version}`)
}

function handleAPICall(request, response) {
  try {
    assertProtocolVersion(request.body.version)
  } catch (error) {
    if (error instanceof ProtocolError) {
      return error.toAPI(response)
    }
  }
}

class InternalError extends createNamespaceError('CRI', { logger: logger }) {
  log(level = 'critical') {
    this.metadata.logger.log({ level, code: this.code, message: this.message })
  }
}

function assertAnchorService(cas) {
  return assertAs(cas instanceof CeramicAnchorService, InternalError, 123, 'Invalid anchor service provided')
}

function checkConfig(config) {
  try {
    assertAnchorService(config.cas)
    ...
  } catch (err) {
    if (error instanceof InternalError) {
      error.log()
    }
  }
}
```

The goal with this library is really to have the minimal set of shared behavior abstracting errors, while being very flexible so it can accommodate the specific needs of our different libraries.